Security.md

```diff
@@ -108,13 +108,76 @@ in-order execution and [post-MVP atomic memory primitives](PostMVP.md#threads).
+to each set, and the instrumentation checks whether the current call target is
+a member of the expected call target set. If this check succeeds, then the
+original call is allowed to proceed, otherwise a failure handler is executed,
+which typically terminates the program.
+
+In WebAssembly, the execution semantics implicitly guarantee the safety of (1)
+through usage of explicit function section indexes, and (3) through a shadow
+stack. Additionally, the type signature of indirect function calls is already
+checked at runtime, effectively implementing coarse-grained type-based control-
+flow integrity for (2). All of this is achieved without explicit runtime
+instrumentation in the module. However, as discussed
+[previously](#memory-safety), this protection does not prevent code reuse
+attacks with function-level granularity against indirect calls. As a result,
+the WebAssembly also supports fine-grained control-flow integrity for (2). This
+leverages built-in support for [forward-edge control-flow integrity][] in the
+Clang/LLVM compiler suite, which has been extended to WebAssembly.
+
+Support for fine-grained control-flow integrity in WebAssembly operates as
+follows. First, the Clang frontend emits type signatures for each function
+and each call site in the LLVM IR. This is used to partition all accessible
+functions in the module into separate expected call target sets based on type.
+Next, instrumentation is inserted at each indirect call site to verify that the
+computed index has the correct type. Currently, a transformation pass
+(`LowerTypeTests`) performs this operation, by using a monotonic counter to
+assign consecutive indexes to each potential call target by type, allowing the
+type check to be implemented efficiently using a integer range comparison. A
+special assembler primitive is used to ensure that these pre-assigned indexes
+are respected by the linker. In the future, this operation will be optimized to
+have no performance penalty (compared to runtime instrumentation), by generating
+[unique indirect tables](Modules.md#table-index-space) for each signature type
+that corresponds to a call target set. This replaces both runtime
+instrumentation and runtime type signature checks with a load-time table
+integrity check for signature type homogeneity.
+
+Enabling fine-grained control-flow integrity (by passing `-fsanitize=cfi` to
+emscripten) has a number of advantages over the default WebAssembly
+configuration. Not only does this better defend against code reuse attacks that
+leverage indirect function calls (2), but it also enhances the built-in function
+signature checks by operating at the C/C++ type level, which is semantically
+richer that the WebAssembly [type level](AstSemantics.md#types), which consists
+of only four value types. Additionally, this feature has virtually no
+performance penalty, by leveraging built-in support for
+[multiple indirect tables](Modules.md#table-index-space) in WebAssembly.
 
```

Comment:
It limits access to functions that can be called indirectly, by restricting usable indexes to those in the selected table. #682 adds support for `call_indirect` to specify a specific table to use.

There are no restrictions on which callers can use various indirect tables.